### DISCUSSION TITLE
Halo2 怎么学 ？

### DISCUSSION THREAD

**Discussion by Demian101** [z2o-k7e/halo2-step-by-step#22] [Q&A]
Q: Halo2 怎么学 ？

@Dream：

使用cario 或者 circom写电路， 是接触不到具体底层电路的设计，优化的，直接裸写电路 成长会很快

Q: 啥叫裸写电路，比如用halo2 原生接口写？

@Dream：
- 对，方便优化， 可以知晓底层生成polynomials ，各种计算开销等
- 我觉得常见的basic 电路（比较大小， lookup等）， 复杂点的如签名电路， merkle tree电路等

嗯嗯。也正准备看一下halo2 原代码的，大佬有没有什么这方面的建议

@Dream：大概挺久以前我看过一遍， 没啥建议，看完并理解文档 halo2 book 后 ，可以看代码了

**Comment by Demian101** [comment]
如题，大家可以在本 discussions 下面留下自己的建议~

**Reply by qizhou** [reply]
我建议可以把halo2本身的例子代码玩一遍，尤其是PSE里面加了很多例子代码 https://github.com/privacy-scaling-explorations/halo2/tree/main/halo2_proofs/examples

**Comment by readygo67** [comment]
我的理解，halo2的编程模式采用 “配置 ->(计算+存储)->验证”三个阶段。
1 . 在chip configure阶段定义约束关系，即在meta.create_gate中从table 中query cell，并将query出的结果(即Expression，expression可以简单的理解为 从某个Cell中获取值的方式，在configure 阶段并不知道具体的值，在验证阶段才会取值)形成约束。
2. 在电路的synthesize 阶段，按照约束关系计算，并将计算结果填入适当的Cell。相当于在table上留下计算的trace。
3.  在电路的verify阶段，从cell 中取出具体的值，并验证约束关系是否成立。
